Separate builders for build tools, drop some unused library dependencies

featheredtoast opened the pull request:
Separate out builders for thpoff, jemalloc and oxipng/jhead
Copy over the results of all libraries to reduce dependencies
for base image needing build tools.

drop extra -dev packages from base image as we only need the nginx runtime libs.

refactor parallel stop without parallel package

mount tmpfs for apt caches
remove unreferenced libtcmalloc

reorganize gem dependencies

featheredtoast committed:
DEV: add separate builders for thpoff, jemalloc, and oxipng

Separate out builders for thpoff, jemalloc and oxipng/jhead
Copy over the results of all libraries to reduce dependencies
for base image needing build tools.

featheredtoast committed:
Drop parallel

Remove the dependency on parallel for two commands

featheredtoast committed:
mount tmpfs for apt caches

featheredtoast committed:
DEV: only install nginx runtime packages

drop -dev packages for nginx compilation now that this is separate

featheredtoast committed:
Remove libtcmalloc - this is unreferenced

..and remove extra cmake reference, this is already in a line above

featheredtoast committed:
DEV: organize apt installs

featheredtoast committed:
correct copy libjemalloc.so

Build in arm vs amd64 acts a bit differently - arm builds to libjemalloc.so.2
and symlinks whereas amd64 does not.

build and copy over library directly to /usr/lib/libjemalloc.so in both cases.

andrewschleifer commented on image/base/sbin/boot:
```diff
@@ -5,15 +5,18 @@
+  for pid in $ORPHANS; do
+    (timeout 5 /bin/bash -c "kill $pid && wait $pid" 2>/dev/null || kill -9 $pid 2>/dev/null) &
+  done
+  wait
```
Do you know what the history is with this? Why are we killing ORPHANS in the first place? Unclean shutdown feels out of scope for this script and ought to be handled one level up.

featheredtoast replied:
I think it's related to how docker/launcher interacted with the containers previously, but I don't have history other than the note in git history:

> FIX: force all sub processes to stop on TERM, wait for them to finish
> This works around docker bug where containers will not stop

Change here attempted to remove parallel without rocking the boat too much on lifecycle.

andrewschleifer commented on image/base/Dockerfile:
```diff
@@ -55,23 +70,34 @@ RUN echo 2.0.`date +%Y%m%d` > /VERSION
 
 RUN --mount=type=tmpfs,target=/var/log \
+  --mount=type=tmpfs,target=/var/cache/apt \
+  --mount=type=tmpfs,target=/var/lib/apt \
```
I don't know what the lifecycle of this container is. Are we potentially increasing the RAM usage for every Discourse here?

featheredtoast replied:
No, this Dockerfile is only done by the builder for building the base dependencies image. All Discourse images /containers build or run later don't inherit a tmpfs from this.

The idea here is just to have an image that is clear of apt caches on downstream images.

We used to do this back in the day by doing a later `RUN rm-all-caches` and flattening image layers, but tmpfs is the way to achieve this now for directories we don't want saved in a final image.

*Cache mounting for apt is also given as an example on [Docker's docs page](https://docs.docker.com/build/cache/optimize/#use-cache-mounts) as well.

andrewschleifer commented on image/base/Dockerfile:
```diff
@@ -55,23 +70,34 @@ RUN echo 2.0.`date +%Y%m%d` > /VERSION
+    gawk anacron wget \
+    psmisc whois brotli \
     pngcrush pngquant ripgrep poppler-utils \
     catdoc antiword \
```
Pulling and labelling the sections below is awesome. Should we do the same with the preceding as `# discourse runtime requirements`?

featheredtoast replied:
I've added it - Just for full transparency, I initially didn't label the first block above as it's less clear to me as to exactly what is a hard discourse requirement (I'm not so sure about a Discourse hard requirement on fping, gawk, psmisc, ripgrep, or whois for example).
Some are known for container boot (anacron, wget, gnupg, sudo, curl).
Others I know are Discourse requirements - pngcrush, pngquant, poppler-utils, catdoc, antiword.

It's possible there's crossover.

andrewschleifer commented on image/base/Dockerfile:
```diff
@@ -127,24 +153,17 @@ RUN ln -s /usr/local/bin/magick /usr/local/bin/animate &&\
+COPY --from=thpoff_builder /usr/local/sbin/thpoff /usr/local/sbin
+COPY --from=jemalloc_builder /usr/lib/libjemalloc.so /usr/lib
+COPY --from=oxipng_builder /usr/local/bin/jhead /usr/local/bin
+COPY --from=oxipng_builder /usr/local/bin/oxipng /usr/local/bin
```
Bikeshed: Infrastructure convention is to use hyphens instead of underscores in these names.

featheredtoast committed:
Update naming convention, prefer dash-names.

Add comment for Discourse dependencies, as best I can put.

featheredtoast commented:
@andrewschleifer updated 👍

featheredtoast committed:
REFACTOR: rename stage name scheme to dashes

updating name to dashes, run ruby lint

featheredtoast committed:
Further split up the build into discourse-runtime-dependencies

Allows for an image to potentially build from discourse-runtime-dependencies

Evict all of the built-time tools from a discourse-runtime-dependencies target

featheredtoast committed:
REFACTORY: rename discourse-dependencies to discourse-build-base

Rename new target to discourse-runtime-base to better reflect.

featheredtoast committed:
Move node, yarn, pnpm to build dependencies stage

node/yarn/pnpm are build-time dependencies.
